File: flpz_code/phononCoupling/transformCell_Map.m
function Mapping=transformCell_Map()


ORIGINALCELL

TARGETCELL


%make lists of all the positions corresponding to each element type for the
%Gamma cell
Gamma_posfrac_Ca=[Gamma_posfrac(1:1,:)];
Gamma_posfrac_Ti=[Gamma_posfrac(2:2,:)];
Gamma_posfrac_O=[Gamma_posfrac(3:5,:)];

%store those in a cell array
Gamma_posCell={Gamma_posfrac_Ca,Gamma_posfrac_Ti,Gamma_posfrac_O};


%make lists of all the positions corresponding to each element type for the
%target cell
Target_posfrac_Ca=[Target_posfrac(1:2,:)];
Target_posfrac_Ti=[Target_posfrac(2:4,:)];
Target_posfrac_O=[Target_posfrac(5:10,:)];

Target_posCell={Target_posfrac_Ca,Target_posfrac_Ti,Target_posfrac_O};



%solving T*P6mmmCell=R3barmCell
%T=R3barmCell*inv(P6mmmCell)

%I think what I want to do is do everything in cartesian, expand the cell,
%then manually find things in it.

%make supercell that extends from -3 to 3 in each direction. overkill but
%should definitely be enough

atomSuperCell={};
aTest=[];
dimSize=3;
for aType=1:length(Gamma_posfrac)%(Gamma_posCell)
    madeSuper=[];
    %[L,~]=size(Gamma_posCell{aType});
    xStencil=[1,0,0];%[ones(L,1),zeros(L,2)];
    yStencil=[0,1,0];%[zeros(L,1),ones(L,1),zeros(L,1)];
    zStencil=[0,0,1];%[zeros(L,2),ones(L,1)];
    for xdim=-dimSize:dimSize
        for ydim=-dimSize:dimSize
            for zdim=-dimSize:dimSize
                shiftedPos=Gamma_posfrac(aType,:)+xdim*xStencil+ydim*yStencil+zdim*zStencil;
                madeSuper=[madeSuper;shiftedPos];
            end
        end
    end
    atomSuperCell{aType}=madeSuper;
    aTest=[aTest;madeSuper];
end


aTest=aTest*GammaCell;



Mapping=zeros(length(Target_posfrac),1);
outDiff=zeros(size(Target_posfrac));

finalDiff=[];
tolerance=.2; %angstroms, tolerance to match atoms
for aType=1:length(Gamma_posfrac)
    thisAtom_orig=atomSuperCell{aType};
    [L,~]=size(thisAtom_orig);
    [Lt,~]=size(Target_posfrac);
    
    for a=1:L
        atomVec=(thisAtom_orig(a,:)*GammaCell);%*rotMat;
        for at=1:Lt
            atomTargetVec=(Target_posfrac(at,:)*TargetCell);
            norm(atomVec-atomTargetVec);
            if norm(atomVec-atomTargetVec)<tolerance
                Mapping(at)=aType;
                disp(['atom type ', num2str(aType),' matches in supercell atom ',num2str(at)])
                outDiff(at,:)=(atomVec-atomTargetVec);
                %outDiff=[outDiff;(atomVec-atomTargetVec)];
            end
        end
    end

    %finalSuperCell=[finalSuperCell;inPos];
    %finalDiff=[finalDiff;outDiff];
end
Mapping

File: flpz_code/phononCoupling/transformCell_Map.py
import numpy as np

ORIGINALCELL

TARGETCELL

def transformCell_Map():

    # Make lists of all the positions corresponding to each element type for the Gamma cell
    Gamma_posfrac_Ca = ORIGINAL_posfrac[0:1, :]
    Gamma_posfrac_Ti = ORIGINAL_posfrac[1:2, :]
    Gamma_posfrac_O = ORIGINAL_posfrac[2:5, :]

    # Store those in a list
    Gamma_posCell = [Gamma_posfrac_Ca, Gamma_posfrac_Ti, Gamma_posfrac_O]

    # Make lists of all the positions corresponding to each element type for the target cell
    Target_posfrac_Ca = TARGET_posfrac[0:2, :]
    Target_posfrac_Ti = TARGET_posfrac[2:4, :]
    Target_posfrac_O = TARGET_posfrac[4:10, :]

    Target_posCell = [Target_posfrac_Ca, Target_posfrac_Ti, Target_posfrac_O]

    # Make supercell that extends from -3 to 3 in each direction
    atomSuperCell = []
    aTest = []
    dimSize = 3
    for aType in range(len(ORIGINAL_posfrac)):
        madeSuper = []
        xStencil = np.array([1, 0, 0])
        yStencil = np.array([0, 1, 0])
        zStencil = np.array([0, 0, 1])
        for xdim in range(-dimSize, dimSize + 1):
            for ydim in range(-dimSize, dimSize + 1):
                for zdim in range(-dimSize, dimSize + 1):
                    shiftedPos = ORIGINAL_posfrac[aType] + xdim * xStencil + ydim * yStencil + zdim * zStencil
                    madeSuper.append(shiftedPos)
        atomSuperCell.append(np.array(madeSuper))
        aTest.extend(madeSuper)

    aTest = np.dot(np.array(aTest), ORIGINALCell)

    Mapping = np.zeros(len(TARGET_posfrac), dtype=int)
    outDiff = np.zeros_like(TARGET_posfrac)

    tolerance = 0.2  # angstroms, tolerance to match atoms
    for aType in range(len(ORIGINAL_posfrac)):
        thisAtom_orig = atomSuperCell[aType]
        for a, atomVec in enumerate(np.dot(thisAtom_orig, ORIGINALCell)):
            for at, atomTargetVec in enumerate(np.dot(TARGET_posfrac, TARGETCell)):
                if np.linalg.norm(atomVec - atomTargetVec) < tolerance:
                    Mapping[at] = aType + 1  # +1 because Python is 0-indexed
                    print(f"atom type {aType + 1} matches in supercell atom {at + 1}")
                    outDiff[at] = atomVec - atomTargetVec

    return Mapping, outDiff

# Call the function and print the results
Mapping, outDiff = transformCell_Map()
print("Mapping:")
print(Mapping)

# I'm not sure if the outdiff works yet
print("\noutDiff:")
print(outDiff)

# Example usage:
# ORIGINALCELL and TARGETCELL should be defined before calling the function
# Gamma_posfrac and Target_posfrac should also be defined
# Mapping, outDiff = transformCell_Map(ORIGINALCELL, TARGETCELL)
